File: ai-services/tracing/tracer.py
"""
iSECTECH Python OpenTelemetry Instrumentation
Production-grade distributed tracing for AI services
"""

import os
import logging
from typing import Dict, Any, Optional
from contextlib import contextmanager

from opentelemetry import trace, metrics
from opentelemetry.exporter.jaeger.thrift import JaegerExporter
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.exporter.otlp.proto.http.metric_exporter import OTLPMetricExporter
from opentelemetry.sdk.trace import TracerProvider, Span
from opentelemetry.sdk.trace.export import BatchSpanProcessor, ConsoleSpanExporter
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.resources import Resource
from opentelemetry.semconv.resource import ResourceAttributes
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
from opentelemetry.baggage.propagation import W3CBaggagePropagator
from opentelemetry.propagators.composite import CompositeHTTPPropagator
from opentelemetry.sdk.trace.sampling import (
    TraceIdRatioBasedSampler,
    AlwaysSample,
    ParentBasedSampler,
)
from opentelemetry.instrumentation.auto_instrumentation import sitecustomize

# ═══════════════════════════════════════════════════════════════════════════════
# CONFIGURATION
# ═══════════════════════════════════════════════════════════════════════════════

SERVICE_NAME = "isectech-ai-services"
SERVICE_NAMESPACE = "isectech"
DEFAULT_VERSION = "1.0.0"

logger = logging.getLogger(__name__)


class TracingConfig:
    """Configuration class for OpenTelemetry tracing"""
    
    def __init__(self):
        self.service_name = os.getenv("OTEL_SERVICE_NAME", SERVICE_NAME)
        self.service_version = os.getenv("OTEL_SERVICE_VERSION", DEFAULT_VERSION)
        self.environment = os.getenv("OTEL_ENVIRONMENT", "development")
        self.otlp_endpoint = os.getenv(
            "OTEL_EXPORTER_OTLP_TRACES_ENDPOINT", 
            "http://localhost:4318/v1/traces"
        )
        self.metrics_endpoint = os.getenv(
            "OTEL_EXPORTER_OTLP_METRICS_ENDPOINT",
            "http://localhost:4318/v1/metrics"
        )
        self.jaeger_endpoint = os.getenv(
            "JAEGER_ENDPOINT", 
            "http://localhost:14268/api/traces"
        )
        self.sampling_rate = float(os.getenv("OTEL_TRACES_SAMPLER_ARG", "0.1"))
        self.debug = os.getenv("OTEL_DEBUG", "false").lower() == "true"
        self.enable_jaeger = os.getenv("ENABLE_JAEGER_EXPORT", "false").lower() == "true"
        self.enable_otlp = os.getenv("ENABLE_OTLP_EXPORT", "true").lower() == "true"
        self.auth_token = os.getenv("OTEL_AUTH_TOKEN", "")


class ISECTECHTracerProvider:
    """Custom tracer provider for iSECTECH AI services"""
    
    def __init__(self, config: TracingConfig):
        self.config = config
        self.tracer_provider = None
        self.tracer = None
        self.meter_provider = None
        self.meter = None
        self._initialize()
    
    def _create_resource(self) -> Resource:
        """Create OpenTelemetry resource with service information"""
        attributes = {
            ResourceAttributes.SERVICE_NAME: self.config.service_name,
            ResourceAttributes.SERVICE_VERSION: self.config.service_version,
            ResourceAttributes.SERVICE_NAMESPACE: SERVICE_NAMESPACE,
            ResourceAttributes.DEPLOYMENT_ENVIRONMENT: self.config.environment,
            ResourceAttributes.SERVICE_INSTANCE_ID: os.getenv("HOSTNAME", "unknown"),
            ResourceAttributes.CONTAINER_NAME: os.getenv("CONTAINER_NAME", "ai-services"),
            ResourceAttributes.CONTAINER_ID: os.getenv("CONTAINER_ID", "unknown"),
            ResourceAttributes.K8S_POD_NAME: os.getenv("K8S_POD_NAME", "unknown"),
            ResourceAttributes.K8S_NAMESPACE_NAME: os.getenv("K8S_NAMESPACE", "default"),
            ResourceAttributes.K8S_CLUSTER_NAME: os.getenv("K8S_CLUSTER_NAME", "isectech-cluster"),
            
            # Custom iSECTECH attributes
            "isectech.component": "ai-services",
            "isectech.team": "ai-team",
            "isectech.environment.type": self.config.environment,
            "isectech.version": self.config.service_version,
        }
        
        return Resource.create(attributes)
    
    def _create_sampler(self):
        """Create trace sampler based on environment"""
        if self.config.environment == "development" or self.config.debug:
            return AlwaysSample()
        
        # Production sampling with parent-based sampling
        return ParentBasedSampler(
            root=TraceIdRatioBasedSampler(self.config.sampling_rate)
        )
    
    def _create_span_processors(self):
        """Create span processors with configured exporters"""
        span_processors = []
        
        # OTLP HTTP Exporter
        if self.config.enable_otlp:
            headers = {}
            if self.config.auth_token:
                headers = {
                    "Authorization": f"Bearer {self.config.auth_token}",
                    "X-Service-Name": self.config.service_name,
                }
            
            otlp_exporter = OTLPSpanExporter(
                endpoint=self.config.otlp_endpoint,
                headers=headers,
                compression="gzip",
            )
            
            span_processors.append(
                BatchSpanProcessor(
                    otlp_exporter,
                    max_queue_size=2048,
                    max_export_batch_size=512,
                    export_timeout_millis=30000,
                    schedule_delay_millis=5000,
                )
            )
        
        # Jaeger Exporter
        if self.config.enable_jaeger:
            jaeger_exporter = JaegerExporter(
                collector_endpoint=self.config.jaeger_endpoint,
            )
            span_processors.append(BatchSpanProcessor(jaeger_exporter))
        
        # Console exporter for development
        if self.config.debug and self.config.environment == "development":
            span_processors.append(BatchSpanProcessor(ConsoleSpanExporter()))
        
        return span_processors
    
    def _create_metric_reader(self):
        """Create metric reader for OTLP metrics export"""
        if not self.config.enable_otlp:
            return None
            
        headers = {}
        if self.config.auth_token:
            headers = {
                "Authorization": f"Bearer {self.config.auth_token}",
                "X-Service-Name": self.config.service_name,
            }
        
        metric_exporter = OTLPMetricExporter(
            endpoint=self.config.metrics_endpoint,
            headers=headers,
            compression="gzip",
        )
        
        return PeriodicExportingMetricReader(
            exporter=metric_exporter,
            export_interval_millis=30000,  # 30 seconds
            export_timeout_millis=5000,    # 5 seconds
        )
    
    def _initialize(self):
        """Initialize the tracer and meter providers"""
        try:
            # Create resource
            resource = self._create_resource()
            
            # Initialize tracer provider
            self.tracer_provider = TracerProvider(
                resource=resource,
                sampler=self._create_sampler(),
            )
            
            # Add span processors
            for processor in self._create_span_processors():
                self.tracer_provider.add_span_processor(processor)
            
            # Set global tracer provider
            trace.set_tracer_provider(self.tracer_provider)
            
            # Create tracer
            self.tracer = trace.get_tracer(
                self.config.service_name,
                self.config.service_version,
            )
            
            # Initialize meter provider
            metric_reader = self._create_metric_reader()
            if metric_reader:
                self.meter_provider = MeterProvider(
                    resource=resource,
                    metric_readers=[metric_reader],
                )
                metrics.set_meter_provider(self.meter_provider)
                self.meter = metrics.get_meter(
                    self.config.service_name,
                    self.config.service_version,
                )
            
            # Set global propagator
            trace.set_global_textmap(
                CompositeHTTPPropagator([
                    TraceContextTextMapPropagator(),
                    W3CBaggagePropagator(),
                ])
            )
            
            logger.info(
                f"✅ OpenTelemetry tracing initialized for {self.config.service_name}",
                extra={
                    "service": self.config.service_name,
                    "version": self.config.service_version,
                    "environment": self.config.environment,
                    "otlp_endpoint": self.config.otlp_endpoint,
                    "sampling_rate": self.config.sampling_rate,
                }
            )
            
        except Exception as e:
            logger.error(f"❌ Failed to initialize OpenTelemetry tracing: {e}")
            raise
    
    def get_tracer(self) -> trace.Tracer:
        """Get the configured tracer"""
        return self.tracer
    
    def get_meter(self) -> Optional[metrics.Meter]:
        """Get the configured meter"""
        return self.meter
    
    def shutdown(self):
        """Shutdown the tracer provider"""
        if self.tracer_provider:
            self.tracer_provider.shutdown()
        if self.meter_provider:
            self.meter_provider.shutdown()


# ═══════════════════════════════════════════════════════════════════════════════
# GLOBAL TRACER PROVIDER
# ═══════════════════════════════════════════════════════════════════════════════

_global_tracer_provider: Optional[ISECTECHTracerProvider] = None


def initialize_tracing(config: Optional[TracingConfig] = None) -> ISECTECHTracerProvider:
    """Initialize global tracing"""
    global _global_tracer_provider
    
    if _global_tracer_provider:
        logger.warning("Tracing already initialized")
        return _global_tracer_provider
    
    if config is None:
        config = TracingConfig()
    
    _global_tracer_provider = ISECTECHTracerProvider(config)
    return _global_tracer_provider


def get_tracer() -> trace.Tracer:
    """Get the global tracer"""
    if _global_tracer_provider:
        return _global_tracer_provider.get_tracer()
    return trace.get_tracer(SERVICE_NAME)


def get_meter() -> Optional[metrics.Meter]:
    """Get the global meter"""
    if _global_tracer_provider:
        return _global_tracer_provider.get_meter()
    return None


def shutdown_tracing():
    """Shutdown global tracing"""
    global _global_tracer_provider
    if _global_tracer_provider:
        _global_tracer_provider.shutdown()
        _global_tracer_provider = None


# ═══════════════════════════════════════════════════════════════════════════════
# CUSTOM SPAN UTILITIES FOR ISECTECH
# ═══════════════════════════════════════════════════════════════════════════════

@contextmanager
def create_security_span(name: str, attributes: Optional[Dict[str, Any]] = None):
    """Create a span for security operations"""
    tracer = get_tracer()
    span_attributes = {
        "isectech.operation_type": "security",
        "isectech.component": "security",
    }
    if attributes:
        span_attributes.update(attributes)
    
    with tracer.start_as_current_span(name, attributes=span_attributes) as span:
        yield span


@contextmanager
def create_ai_span(name: str, model_name: str = None, attributes: Optional[Dict[str, Any]] = None):
    """Create a span for AI/ML operations"""
    tracer = get_tracer()
    span_attributes = {
        "isectech.operation_type": "ai",
        "isectech.component": "ai",
    }
    if model_name:
        span_attributes["ai.model.name"] = model_name
    if attributes:
        span_attributes.update(attributes)
    
    with tracer.start_as_current_span(name, attributes=span_attributes) as span:
        yield span


@contextmanager
def create_business_span(name: str, user_id: str = None, attributes: Optional[Dict[str, Any]] = None):
    """Create a span for business operations"""
    tracer = get_tracer()
    span_attributes = {
        "isectech.operation_type": "business",
        "isectech.user_id": user_id or "anonymous",
    }
    if attributes:
        span_attributes.update(attributes)
    
    with tracer.start_as_current_span(name, attributes=span_attributes) as span:
        yield span


@contextmanager
def create_database_span(operation: str, table: str = None, attributes: Optional[Dict[str, Any]] = None):
    """Create a span for database operations"""
    tracer = get_tracer()
    span_name = f"db.{operation}"
    if table:
        span_name += f" {table}"
    
    span_attributes = {
        "isectech.operation_type": "database",
        "db.operation": operation,
        "isectech.component": "database",
    }
    if table:
        span_attributes["db.sql.table"] = table
    if attributes:
        span_attributes.update(attributes)
    
    with tracer.start_as_current_span(span_name, attributes=span_attributes) as span:
        yield span


# ═══════════════════════════════════════════════════════════════════════════════
# SPAN UTILITIES
# ═══════════════════════════════════════════════════════════════════════════════

def add_security_event(event_type: str, attributes: Optional[Dict[str, Any]] = None):
    """Add a security event to the current span"""
    span = trace.get_current_span()
    if span.is_recording():
        event_attributes = {"isectech.security.event_type": event_type}
        if attributes:
            event_attributes.update(attributes)
        span.add_event("security.event", event_attributes)


def add_error_event(error: Exception, error_type: str = None):
    """Add an error event to the current span"""
    span = trace.get_current_span()
    if span.is_recording():
        span.record_exception(error)
        span.set_attributes({
            "isectech.error": True,
            "isectech.error.type": error_type or type(error).__name__,
        })
        span.set_status(trace.Status(trace.StatusCode.ERROR, str(error)))


def get_trace_id() -> Optional[str]:
    """Get the current trace ID"""
    span = trace.get_current_span()
    if span.get_span_context().is_valid:
        return format(span.get_span_context().trace_id, "032x")
    return None


def get_span_id() -> Optional[str]:
    """Get the current span ID"""
    span = trace.get_current_span()
    if span.get_span_context().is_valid:
        return format(span.get_span_context().span_id, "016x")
    return None


def add_span_attributes(attributes: Dict[str, Any]):
    """Add attributes to the current span"""
    span = trace.get_current_span()
    if span.is_recording():
        span.set_attributes(attributes)


# ═══════════════════════════════════════════════════════════════════════════════
# AI/ML SPECIFIC UTILITIES
# ═══════════════════════════════════════════════════════════════════════════════

def trace_model_inference(model_name: str, input_size: int = None, output_size: int = None):
    """Decorator for tracing AI model inference"""
    def decorator(func):
        def wrapper(*args, **kwargs):
            with create_ai_span(
                f"ai.inference.{model_name}",
                model_name=model_name,
                attributes={
                    "ai.model.input_size": input_size,
                    "ai.model.output_size": output_size,
                }
            ) as span:
                try:
                    result = func(*args, **kwargs)
                    span.set_attribute("ai.inference.success", True)
                    return result
                except Exception as e:
                    span.set_attribute("ai.inference.success", False)
                    add_error_event(e, "ai_inference_error")
                    raise
        return wrapper
    return decorator


def trace_data_processing(operation_type: str):
    """Decorator for tracing data processing operations"""
    def decorator(func):
        def wrapper(*args, **kwargs):
            with create_ai_span(
                f"data.processing.{operation_type}",
                attributes={"data.operation_type": operation_type}
            ) as span:
                try:
                    result = func(*args, **kwargs)
                    span.set_attribute("data.processing.success", True)
                    return result
                except Exception as e:
                    span.set_attribute("data.processing.success", False)
                    add_error_event(e, "data_processing_error")
                    raise
        return wrapper
    return decorator


# ═══════════════════════════════════════════════════════════════════════════════
# AUTO-INITIALIZATION
# ═══════════════════════════════════════════════════════════════════════════════

# Auto-initialize tracing if not in test environment
if (
    os.getenv("PYTEST_CURRENT_TEST") is None and 
    os.getenv("DISABLE_TRACING", "false").lower() != "true"
):
    try:
        initialize_tracing()
    except Exception as e:
        logger.error(f"Failed to auto-initialize tracing: {e}")


# ═══════════════════════════════════════════════════════════════════════════════
# CLEANUP ON EXIT
# ═══════════════════════════════════════════════════════════════════════════════

import atexit
atexit.register(shutdown_tracing)

File: backend/tracing/tracer.go
// iSECTECH Go OpenTelemetry Instrumentation
// Production-grade distributed tracing for Go backend services

package tracing

import (
	"context"
	"fmt"
	"os"
	"strconv"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/jaeger"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	"go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.20.0"
	oteltrace "go.opentelemetry.io/otel/trace"
)

// ═══════════════════════════════════════════════════════════════════════════════
// CONFIGURATION CONSTANTS
// ═══════════════════════════════════════════════════════════════════════════════

const (
	ServiceName      = "isectech-backend"
	ServiceNamespace = "isectech"
	DefaultVersion   = "1.0.0"
)

// TracingConfig holds the configuration for OpenTelemetry tracing
type TracingConfig struct {
	ServiceName     string
	ServiceVersion  string
	Environment     string
	OTLPEndpoint    string
	JaegerEndpoint  string
	SamplingRate    float64
	Debug           bool
	EnableJaeger    bool
	EnableOTLP      bool
}

// TracerProvider wraps the OpenTelemetry tracer provider
type TracerProvider struct {
	provider    *trace.TracerProvider
	tracer      oteltrace.Tracer
	config      *TracingConfig
	shutdown    func(context.Context) error
}

// ═══════════════════════════════════════════════════════════════════════════════
// CONFIGURATION BUILDER
// ═══════════════════════════════════════════════════════════════════════════════

// NewTracingConfig creates a new tracing configuration from environment variables
func NewTracingConfig() *TracingConfig {
	config := &TracingConfig{
		ServiceName:     getEnvOrDefault("OTEL_SERVICE_NAME", ServiceName),
		ServiceVersion:  getEnvOrDefault("OTEL_SERVICE_VERSION", DefaultVersion),
		Environment:     getEnvOrDefault("OTEL_ENVIRONMENT", "development"),
		OTLPEndpoint:    getEnvOrDefault("OTEL_EXPORTER_OTLP_TRACES_ENDPOINT", "http://localhost:4318/v1/traces"),
		JaegerEndpoint:  getEnvOrDefault("JAEGER_ENDPOINT", "http://localhost:14268/api/traces"),
		SamplingRate:    getEnvFloatOrDefault("OTEL_TRACES_SAMPLER_ARG", 0.1),
		Debug:           getEnvBoolOrDefault("OTEL_DEBUG", false),
		EnableJaeger:    getEnvBoolOrDefault("ENABLE_JAEGER_EXPORT", false),
		EnableOTLP:      getEnvBoolOrDefault("ENABLE_OTLP_EXPORT", true),
	}
	
	return config
}

// ═══════════════════════════════════════════════════════════════════════════════
// RESOURCE CREATION
// ═══════════════════════════════════════════════════════════════════════════════

// createResource creates an OpenTelemetry resource with service and environment information
func createResource(config *TracingConfig) (*resource.Resource, error) {
	attributes := []attribute.KeyValue{
		semconv.ServiceNameKey.String(config.ServiceName),
		semconv.ServiceVersionKey.String(config.ServiceVersion),
		semconv.ServiceNamespaceKey.String(ServiceNamespace),
		semconv.DeploymentEnvironmentKey.String(config.Environment),
		semconv.ServiceInstanceIDKey.String(getEnvOrDefault("HOSTNAME", "unknown")),
		semconv.ContainerNameKey.String(getEnvOrDefault("CONTAINER_NAME", "backend")),
		semconv.ContainerIDKey.String(getEnvOrDefault("CONTAINER_ID", "unknown")),
		semconv.K8SPodNameKey.String(getEnvOrDefault("K8S_POD_NAME", "unknown")),
		semconv.K8SNamespaceNameKey.String(getEnvOrDefault("K8S_NAMESPACE", "default")),
		semconv.K8SClusterNameKey.String(getEnvOrDefault("K8S_CLUSTER_NAME", "isectech-cluster")),
		
		// Custom iSECTECH attributes
		attribute.String("isectech.component", "backend"),
		attribute.String("isectech.team", "backend-team"),
		attribute.String("isectech.environment.type", config.Environment),
		attribute.String("isectech.version", config.ServiceVersion),
	}

	return resource.NewWithAttributes(
		semconv.SchemaURL,
		attributes...,
	)
}

// ═══════════════════════════════════════════════════════════════════════════════
// EXPORTER CREATION
// ═══════════════════════════════════════════════════════════════════════════════

// createOTLPExporter creates an OTLP HTTP trace exporter
func createOTLPExporter(ctx context.Context, config *TracingConfig) (trace.SpanExporter, error) {
	options := []otlptracehttp.Option{
		otlptracehttp.WithEndpoint(config.OTLPEndpoint),
		otlptracehttp.WithInsecure(), // Use HTTPS in production
		otlptracehttp.WithCompression(otlptracehttp.GzipCompression),
	}

	// Add authentication headers if token is provided
	if token := os.Getenv("OTEL_AUTH_TOKEN"); token != "" {
		headers := map[string]string{
			"Authorization":    fmt.Sprintf("Bearer %s", token),
			"X-Service-Name":   config.ServiceName,
		}
		options = append(options, otlptracehttp.WithHeaders(headers))
	}

	return otlptracehttp.New(ctx, options...)
}

// createJaegerExporter creates a Jaeger trace exporter
func createJaegerExporter(config *TracingConfig) (trace.SpanExporter, error) {
	return jaeger.New(jaeger.WithCollectorEndpoint(
		jaeger.WithEndpoint(config.JaegerEndpoint),
	))
}

// ═══════════════════════════════════════════════════════════════════════════════
// SAMPLER CREATION
// ═══════════════════════════════════════════════════════════════════════════════

// createSampler creates a trace sampler based on environment and configuration
func createSampler(config *TracingConfig) trace.Sampler {
	if config.Environment == "development" || config.Debug {
		return trace.AlwaysSample()
	}

	// Production sampling with parent-based sampling
	return trace.ParentBased(trace.TraceIDRatioBased(config.SamplingRate))
}

// ═══════════════════════════════════════════════════════════════════════════════
// TRACER PROVIDER INITIALIZATION
// ═══════════════════════════════════════════════════════════════════════════════

// NewTracerProvider creates and initializes a new OpenTelemetry tracer provider
func NewTracerProvider(ctx context.Context, config *TracingConfig) (*TracerProvider, error) {
	// Create resource
	res, err := createResource(config)
	if err != nil {
		return nil, fmt.Errorf("failed to create resource: %w", err)
	}

	// Create span processors with exporters
	var spanProcessors []trace.SpanProcessor

	// OTLP Exporter
	if config.EnableOTLP {
		otlpExporter, err := createOTLPExporter(ctx, config)
		if err != nil {
			return nil, fmt.Errorf("failed to create OTLP exporter: %w", err)
		}
		spanProcessors = append(spanProcessors, trace.NewBatchSpanProcessor(otlpExporter,
			trace.WithBatchTimeout(5*time.Second),
			trace.WithMaxExportBatchSize(512),
			trace.WithMaxQueueSize(2048),
		))
	}

	// Jaeger Exporter
	if config.EnableJaeger {
		jaegerExporter, err := createJaegerExporter(config)
		if err != nil {
			return nil, fmt.Errorf("failed to create Jaeger exporter: %w", err)
		}
		spanProcessors = append(spanProcessors, trace.NewBatchSpanProcessor(jaegerExporter))
	}

	// Create tracer provider
	provider := trace.NewTracerProvider(
		trace.WithResource(res),
		trace.WithSampler(createSampler(config)),
	)

	// Add span processors
	for _, processor := range spanProcessors {
		provider.RegisterSpanProcessor(processor)
	}

	// Set global tracer provider
	otel.SetTracerProvider(provider)

	// Set global text map propagator for trace context propagation
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	// Create tracer
	tracer := provider.Tracer(
		config.ServiceName,
		oteltrace.WithInstrumentationVersion(config.ServiceVersion),
		oteltrace.WithSchemaURL(semconv.SchemaURL),
	)

	return &TracerProvider{
		provider: provider,
		tracer:   tracer,
		config:   config,
		shutdown: provider.Shutdown,
	}, nil
}

// ═══════════════════════════════════════════════════════════════════════════════
// TRACER PROVIDER METHODS
// ═══════════════════════════════════════════════════════════════════════════════

// Tracer returns the OpenTelemetry tracer
func (tp *TracerProvider) Tracer() oteltrace.Tracer {
	return tp.tracer
}

// Shutdown gracefully shuts down the tracer provider
func (tp *TracerProvider) Shutdown(ctx context.Context) error {
	if tp.shutdown != nil {
		return tp.shutdown(ctx)
	}
	return nil
}

// ═══════════════════════════════════════════════════════════════════════════════
// CUSTOM SPAN UTILITIES FOR ISECTECH
// ═══════════════════════════════════════════════════════════════════════════════

// StartSecuritySpan starts a new span for security operations
func (tp *TracerProvider) StartSecuritySpan(ctx context.Context, name string, opts ...oteltrace.SpanStartOption) (context.Context, oteltrace.Span) {
	opts = append(opts, oteltrace.WithAttributes(
		attribute.String("isectech.operation_type", "security"),
		attribute.String("isectech.component", "security"),
	))
	return tp.tracer.Start(ctx, name, opts...)
}

// StartBusinessSpan starts a new span for business operations
func (tp *TracerProvider) StartBusinessSpan(ctx context.Context, name string, userID string, opts ...oteltrace.SpanStartOption) (context.Context, oteltrace.Span) {
	opts = append(opts, oteltrace.WithAttributes(
		attribute.String("isectech.operation_type", "business"),
		attribute.String("isectech.user_id", userID),
	))
	return tp.tracer.Start(ctx, name, opts...)
}

// StartDatabaseSpan starts a new span for database operations
func (tp *TracerProvider) StartDatabaseSpan(ctx context.Context, operation string, table string, opts ...oteltrace.SpanStartOption) (context.Context, oteltrace.Span) {
	opts = append(opts, oteltrace.WithAttributes(
		attribute.String("isectech.operation_type", "database"),
		attribute.String("db.operation", operation),
		attribute.String("db.sql.table", table),
		attribute.String("isectech.component", "database"),
	))
	return tp.tracer.Start(ctx, fmt.Sprintf("db.%s %s", operation, table), opts...)
}

// StartHTTPSpan starts a new span for HTTP operations
func (tp *TracerProvider) StartHTTPSpan(ctx context.Context, method string, url string, opts ...oteltrace.SpanStartOption) (context.Context, oteltrace.Span) {
	opts = append(opts, oteltrace.WithAttributes(
		attribute.String("isectech.operation_type", "http"),
		attribute.String("http.method", method),
		attribute.String("http.url", url),
		attribute.String("isectech.component", "http"),
	))
	return tp.tracer.Start(ctx, fmt.Sprintf("%s %s", method, url), opts...)
}

// ═══════════════════════════════════════════════════════════════════════════════
// SPAN UTILITIES
// ═══════════════════════════════════════════════════════════════════════════════

// AddSecurityEvent adds a security-related event to the current span
func AddSecurityEvent(ctx context.Context, eventType string, attributes map[string]interface{}) {
	span := oteltrace.SpanFromContext(ctx)
	if span.IsRecording() {
		attrs := []attribute.KeyValue{
			attribute.String("isectech.security.event_type", eventType),
		}
		for k, v := range attributes {
			switch val := v.(type) {
			case string:
				attrs = append(attrs, attribute.String(k, val))
			case int:
				attrs = append(attrs, attribute.Int(k, val))
			case bool:
				attrs = append(attrs, attribute.Bool(k, val))
			}
		}
		span.AddEvent("security.event", oteltrace.WithAttributes(attrs...))
	}
}

// AddErrorEvent adds an error event to the current span
func AddErrorEvent(ctx context.Context, err error, errorType string) {
	span := oteltrace.SpanFromContext(ctx)
	if span.IsRecording() {
		span.RecordError(err, oteltrace.WithAttributes(
			attribute.String("isectech.error.type", errorType),
			attribute.Bool("isectech.error", true),
		))
		span.SetStatus(oteltrace.StatusCodeError, err.Error())
	}
}

// GetTraceID returns the trace ID from the current context
func GetTraceID(ctx context.Context) string {
	span := oteltrace.SpanFromContext(ctx)
	if span.SpanContext().IsValid() {
		return span.SpanContext().TraceID().String()
	}
	return ""
}

// GetSpanID returns the span ID from the current context
func GetSpanID(ctx context.Context) string {
	span := oteltrace.SpanFromContext(ctx)
	if span.SpanContext().IsValid() {
		return span.SpanContext().SpanID().String()
	}
	return ""
}

// ═══════════════════════════════════════════════════════════════════════════════
// UTILITY FUNCTIONS
// ═══════════════════════════════════════════════════════════════════════════════

// getEnvOrDefault returns the environment variable value or default if not set
func getEnvOrDefault(key, defaultValue string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}

// getEnvFloatOrDefault returns the environment variable as float64 or default if not set/invalid
func getEnvFloatOrDefault(key string, defaultValue float64) float64 {
	if value := os.Getenv(key); value != "" {
		if parsed, err := strconv.ParseFloat(value, 64); err == nil {
			return parsed
		}
	}
	return defaultValue
}

// getEnvBoolOrDefault returns the environment variable as bool or default if not set/invalid
func getEnvBoolOrDefault(key string, defaultValue bool) bool {
	if value := os.Getenv(key); value != "" {
		if parsed, err := strconv.ParseBool(value); err == nil {
			return parsed
		}
	}
	return defaultValue
}

// ═══════════════════════════════════════════════════════════════════════════════
// GLOBAL TRACER PROVIDER
// ═══════════════════════════════════════════════════════════════════════════════

var globalTracerProvider *TracerProvider

// InitializeGlobalTracer initializes the global tracer provider
func InitializeGlobalTracer(ctx context.Context) error {
	config := NewTracingConfig()
	
	tp, err := NewTracerProvider(ctx, config)
	if err != nil {
		return fmt.Errorf("failed to initialize tracer provider: %w", err)
	}
	
	globalTracerProvider = tp
	
	fmt.Printf("✅ OpenTelemetry tracing initialized for %s\n", config.ServiceName)
	fmt.Printf("   Service: %s v%s\n", config.ServiceName, config.ServiceVersion)
	fmt.Printf("   Environment: %s\n", config.Environment)
	fmt.Printf("   OTLP Endpoint: %s\n", config.OTLPEndpoint)
	fmt.Printf("   Sampling Rate: %.2f\n", config.SamplingRate)
	
	return nil
}

// GetGlobalTracer returns the global tracer
func GetGlobalTracer() oteltrace.Tracer {
	if globalTracerProvider != nil {
		return globalTracerProvider.Tracer()
	}
	return otel.Tracer(ServiceName)
}

// ShutdownGlobalTracer shuts down the global tracer provider
func ShutdownGlobalTracer(ctx context.Context) error {
	if globalTracerProvider != nil {
		return globalTracerProvider.Shutdown(ctx)
	}
	return nil
}